pkg/signature/algorithm_registry.go

```diff
@@ -0,0 +1,239 @@
+//
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signature
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"fmt"
+
+	v1 "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+)
+
+// PublicKeyType represents the public key algorithm for a given signature algorithm.
+type PublicKeyType uint
+
+const (
+	RSA     PublicKeyType = iota // RSA public key
+	ECDSA                        // ECDSA public key
+	ED25519                      // ED25519 public key
+)
+
+// RSAKeySize represents the size of an RSA public key.
+type RSAKeySize int
+
+// AlgorithmDetails exposes relevant information for a given signature algorithm.
+type AlgorithmDetails interface {
+	// GetSignatureAlgorithm returns the algorithm registry.
+	GetSignatureAlgorithm() v1.KnownSignatureAlgorithm
+
+	// GetKeyType returns the public key algorithm for the given signature algorithm.
+	GetKeyType() PublicKeyType
+
+	// GetHashType returns the hash algorithm for a given signature algorithm.
+	GetHashType() crypto.Hash
+
+	// GetRSAKeySize returns the key size if the public key algorithm is RSA.
+	// Otherwise, an error is returned.
+	GetRSAKeySize() (RSAKeySize, error)
+
+	// GetECDSACurve returns the curve if the public key algorithm is ECDSA.
+	// Otherwise, an error is returned.
+	GetECDSACurve() (*elliptic.Curve, error)
+
+	checkKey(crypto.PublicKey) bool
+	checkHash(crypto.Hash) bool
+}
+
+type algorithmDetailsImpl struct {
```

Comment:
Can you add some comments about what each of these fields represent, particularly `flagValue` and `extraKeyParams`?

Reply:
[021f385](https://github.com/sigstore/sigstore/pull/1601/commits/021f385ab7e86b9c2f180e11f266f6855e0b122d)